File: components/elements/text_element.cpp
#include "text_element.h"

namespace esphome::elements {

static const char* const TEXT_ELEMENT_TAG = "elements.text";

using display::TextAlign;

const int TEXT_ALIGN_X = (int)TextAlign::LEFT |
                         (int)TextAlign::CENTER_HORIZONTAL |
                         (int)TextAlign::RIGHT;
const int TEXT_ALIGN_Y = (int)TextAlign::TOP | (int)TextAlign::CENTER_VERTICAL |
                         (int)TextAlign::BASELINE | (int)TextAlign::BOTTOM;

void TextElement::draw(display::Display& display) {
  // Update the text, if necessary.
  std::string new_text = get_text();
  if (text_ != new_text) {
    text_ = new_text;
    request_measurement_ = true;
  }

  // If there is no text, signal completion once and skip drawing.
  if (text_.empty()) {
    if (!complete_) {
      complete_ = true;
      on_complete();
    }
    return;
  }

  // Compute the placement of the text.
  Point<int> point = anchor_.get(Point<int>::from_extent(display));

  // Measure the text, if necessary, using the un-scrolled base point.
  if (request_measurement_) {
    display.get_text_bounds(point.x, point.y, text_.c_str(), font_, align_,
                            &measured_position_.x, &measured_position_.y,
                            &measured_extent_.x, &measured_extent_.y);
    request_measurement_ = false;
  }

  // Update the placement, if we scroll.
  if (scroll_mode_ != ScrollMode::NONE) {
    uint32_t now = millis();
    if (last_scroll_time_ > 0) {
      scroll_offset_ += (now - last_scroll_time_) * scroll_speed_ / 1000.0f;
    }
    last_scroll_time_ = now;
    int dist = static_cast<int>(scroll_offset_);
    Point<int> delta = {
        .x = (scroll_mode_ == ScrollMode::RIGHT_TO_LEFT ? 1 : 0) -
             (scroll_mode_ == ScrollMode::LEFT_TO_RIGHT ? 1 : 0),
        .y = (scroll_mode_ == ScrollMode::TOP_TO_BOTTOM ? 1 : 0) -
             (scroll_mode_ == ScrollMode::BOTTOM_TO_TOP ? 1 : 0),
    };
    point.x += delta.x * dist;
    point.y += delta.y * dist;
    Point<int> current = {
        .x = measured_position_.x + delta.x * dist,
        .y = measured_position_.y + delta.y * dist,
    };
    if (current.x + measured_extent_.x < 0 || current.x > display.get_width() ||
        current.y + measured_extent_.y < 0 ||
        current.y > display.get_height()) {
      scroll_offset_ = 0.0f;
      complete_ = false;
      on_complete();
      return;
    }
  }

  // Draw the text.
  display.print(point.x, point.y, font_, color_, align_, text_.c_str(),
                background_color_);
}

void TextElement::on_show() {
  switch (scroll_mode_) {
    case ScrollMode::NONE:
      break;
    case ScrollMode::LEFT_TO_RIGHT:
      anchor_.fraction.x = 1.0f;
      align_ = TextAlign((int(align_) & ~TEXT_ALIGN_X) | int(TextAlign::LEFT));
      break;
    case ScrollMode::RIGHT_TO_LEFT:
      anchor_.fraction.x = 0.0f;
      align_ = TextAlign((int(align_) & ~TEXT_ALIGN_X) | int(TextAlign::RIGHT));
      break;
    case ScrollMode::BOTTOM_TO_TOP:
      anchor_.fraction.y = 1.0f;
      align_ = TextAlign((int(align_) & ~TEXT_ALIGN_Y) | int(TextAlign::TOP));
      break;
    case ScrollMode::TOP_TO_BOTTOM:
      anchor_.fraction.y = 0.0f;
      align_ =
          TextAlign((int(align_) & ~TEXT_ALIGN_Y) | int(TextAlign::BOTTOM));
      break;
  }
  scroll_offset_ = 0.0f;
  last_scroll_time_ = millis();
  complete_ = false;
  Element::on_show();
}

void TextElement::on_hide() {
  scroll_offset_ = 0.0f;
  last_scroll_time_ = 0;
  complete_ = false;
  Element::on_hide();
}

void TextElement::on_complete() {
  ESP_LOGI(TEXT_ELEMENT_TAG, "Triggering `on_complete` for %s (%p)",
           get_type_name(), this);
  on_complete_callbacks_.call(this);
}

bool TextElement::is_active() { return !get_text().empty(); }

std::string StaticTextElement::get_text() { return text_; }

std::string DynamicTextElement::get_text() { return lambda_(*this); }

std::string TimeTextElement::get_text() {
  return time_->now().strftime(format_);
}

}  // namespace esphome::elements

File: components/elements/text_element.py
import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import font, time
from esphome.const import CONF_ID

from . import shared
from . import element
from . import element_registry
from . import color
from . import geometry

CONF_ALIGN = "align"
CONF_ANCHOR = "anchor"
CONF_BACKGROUND_COLOR = "background_color"
CONF_COLOR = "color"
CONF_FONT = "font"
CONF_FORMAT = "format"
CONF_LAMBDA = "lambda"
CONF_SCROLL_MODE = "scroll_mode"
CONF_SCROLL_SPEED = "scroll_speed"
CONF_TEXT = "text"
CONF_TIME = "time"

# Scroll Mode

ScrollMode = shared.elements_ns.enum("ScrollMode", is_class=True)

SCROLL_MODE = {
    "NONE": ScrollMode.NONE,
    "LEFT_TO_RIGHT": ScrollMode.LEFT_TO_RIGHT,
    "RIGHT_TO_LEFT": ScrollMode.RIGHT_TO_LEFT,
    "BOTTOM_TO_TOP": ScrollMode.BOTTOM_TO_TOP,
    "TOP_TO_BOTTOM": ScrollMode.TOP_TO_BOTTOM,
}

# Text Align

TextAlign = shared.display_ns.enum("TextAlign", is_class=True)

TEXT_ALIGN = {
    "TOP": TextAlign.TOP,
    "CENTER_VERTICAL": TextAlign.CENTER_VERTICAL,
    "BASELINE": TextAlign.BASELINE,
    "BOTTOM": TextAlign.BOTTOM,
    "LEFT": TextAlign.LEFT,
    "CENTER_HORIZONTAL": TextAlign.CENTER_HORIZONTAL,
    "RIGHT": TextAlign.RIGHT,
    "TOP_LEFT": TextAlign.TOP_LEFT,
    "TOP_CENTER": TextAlign.TOP_CENTER,
    "TOP_RIGHT": TextAlign.TOP_RIGHT,
    "CENTER_LEFT": TextAlign.CENTER_LEFT,
    "CENTER": TextAlign.CENTER,
    "CENTER_RIGHT": TextAlign.CENTER_RIGHT,
    "BASELINE_LEFT": TextAlign.BASELINE_LEFT,
    "BASELINE_CENTER": TextAlign.BASELINE_CENTER,
    "BASELINE_RIGHT": TextAlign.BASELINE_RIGHT,
    "BOTTOM_LEFT": TextAlign.BOTTOM_LEFT,
    "BOTTOM_CENTER": TextAlign.BOTTOM_CENTER,
    "BOTTOM_RIGHT": TextAlign.BOTTOM_RIGHT,
}

# Abstract Text Element

TextElement = shared.elements_ns.class_("TextElement", element.Element)

TEXT_ELEMENT_SCHEMA = element.ELEMENT_SCHEMA.extend(
    {
        cv.Required(CONF_FONT): cv.use_id(font.Font),
        cv.Optional(CONF_COLOR, default="#ffffff"): color.COLOR_SCHEMA,
        cv.Optional(CONF_BACKGROUND_COLOR): color.COLOR_SCHEMA,
        cv.Optional(CONF_ANCHOR, default={}): geometry.anchor_schema(),
        cv.Optional(CONF_ALIGN, default="CENTER"): cv.enum(
            TEXT_ALIGN, upper=True, space="_"
        ),
        cv.Optional(CONF_SCROLL_MODE): cv.enum(SCROLL_MODE, upper=True, space="_"),
        cv.Optional(CONF_SCROLL_SPEED): cv.float_range(min=0),
    }
)


async def text_element_to_code(config, component, parent):
    var = await element.element_to_code(config, component, parent)
    if conf := config.get(CONF_FONT):
        value = await cg.get_variable(conf)
        cg.add(var.set_font(value))
    if conf := config.get(CONF_COLOR):
        value = await color.color_to_code(conf)
        cg.add(var.set_color(value))
    if conf := config.get(CONF_BACKGROUND_COLOR):
        value = await color.color_to_code(conf)
        cg.add(var.set_background_color(value))
    if conf := config.get(CONF_ANCHOR):
        value = await geometry.anchor_to_code(conf)
        cg.add(var.set_anchor(value))
    if conf := config.get(CONF_ALIGN):
        cg.add(var.set_align(conf))
    if conf := config.get(CONF_SCROLL_MODE):
        cg.add(var.set_scroll_mode(conf))
    if conf := config.get(CONF_SCROLL_SPEED):
        cg.add(var.set_scroll_speed(conf))
    return var


# Static Text Element

StaticTextElement = shared.elements_ns.class_("StaticTextElement", TextElement)

STATIC_TEXT_ELEMENT_SCHEMA = TEXT_ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(StaticTextElement),
        cv.Optional(CONF_TEXT): cv.string,
    }
)


async def static_text_element_to_code(config, component, parent):
    var = await text_element_to_code(config, component, parent)
    if conf := config.get(CONF_TEXT):
        cg.add(var.set_text(conf))
    return var


element_registry.register_element(
    "static_text", STATIC_TEXT_ELEMENT_SCHEMA, static_text_element_to_code
)

# Dynamic Text Element

DynamicTextElement = shared.elements_ns.class_("DynamicTextElement", TextElement)
DynamicTextRef = DynamicTextElement.operator("ref")

DYNAMIC_TEXT_ELEMENT_SCHEMA = TEXT_ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(DynamicTextElement),
        cv.Required(CONF_LAMBDA): cv.returning_lambda,
    }
)


async def dynamic_text_element_to_code(config, component, parent):
    var = await text_element_to_code(config, component, parent)
    if conf := config.get(CONF_LAMBDA):
        value = await cg.process_lambda(
            conf,
            [(DynamicTextRef, "element")],
            return_type=cg.std_string,
        )
        cg.add(var.set_lambda(value))
    return var


element_registry.register_element(
    "dynamic_text", DYNAMIC_TEXT_ELEMENT_SCHEMA, dynamic_text_element_to_code
)

# Time Text Element

TimeTextElement = shared.elements_ns.class_("TimeTextElement", TextElement)

TIME_TEXT_ELEMENT_SCHEMA = TEXT_ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(TimeTextElement),
        cv.Required(CONF_TIME): cv.use_id(time.RealTimeClock),
        cv.Optional(CONF_FORMAT, default="%H:%M:%S"): cv.string,
    }
)


async def time_text_element_to_code(config, component, parent):
    var = await text_element_to_code(config, component, parent)
    if conf := config.get(CONF_TIME):
        value = await cg.get_variable(conf)
        cg.add(var.set_time(value))
    if conf := config.get(CONF_FORMAT):
        cg.add(var.set_format(conf))
    return var


element_registry.register_element(
    "time_text", TIME_TEXT_ELEMENT_SCHEMA, time_text_element_to_code
)
